### PR TITLE
Add curcuit breaker

### PR DESCRIPTION
**Pull request by nrwiersma**

### CODE REVIEW

File: breaker/breaker.go [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+	return FuseFunc(func(c Counter) bool {
+		return c.ConsecutiveFailures > count
+	})
+}
+
+// RateFuse trips the Breaker when the percentage of failures exceeds the given rate.
+//
+// rate should be between 0 and 100.
+func RateFuse(rate uint64) FuseFunc {
+	return FuseFunc(func(c Counter) bool {
```
Comment (michalkurzeja) [inline breaker/breaker.go]: panic on rate > 100?

File: breaker/breaker.go [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+	for _, opt := range opts {
+		opt(b)
+	}
+
+	return b
+}
+
+// Run runs the given request if the Breaker allows it.
+//
+// Run returns an error immediately is the Breaker rejects the request,
```
Comment (michalkurzeja) [inline breaker/breaker.go]: Typo: `is` => `if`

File: breaker/breaker.go [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+// otherwise it returns the result of the request.
+func (b *Breaker) Run(req func() error) error {
```
Comment (michalkurzeja) [inline breaker/breaker.go]: There are 3 deferred statements executed in the course of this func. This **might** slow down some high-throughput cases. Maybe it's worth benchmarking it with and without defers?

Comment (nrwiersma) [reply breaker/breaker.go]: We can tackle this in a second pass

**Pull request merged**